ledger/participant-state/src/main/scala/com/daml/ledger/participant/state/v2/WriteService.scala

```diff
@@ -0,0 +1,155 @@
+// Copyright (c) 2021 Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package com.daml.ledger.participant.state.v2
+
+import java.util.concurrent.CompletionStage
+
+import com.daml.ledger.api.health.ReportsHealth
+import com.daml.ledger.participant.state.v1.{
+  SubmittedTransaction,
+  TransactionMeta,
+  WriteConfigService,
+  WritePackagesService,
+  WriteParticipantPruningService,
+  WritePartyService
+}
+import com.daml.telemetry.TelemetryContext
+
+/** An interface to change a ledger via a participant.
+  *
+  * The methods in this interface are all methods that are supported
+  * *uniformly* across all ledger participant implementations. Methods for
+  * uploading packages, on-boarding parties, and changing ledger-wide
+  * configuration are specific to a ledger and therefore to a participant
+  * implementation. Moreover, these methods usually require admin-level
+  * privileges, whose granting is also specific to a ledger.
+  *
+  * If a ledger is run for testing only, there is the option for quite freely
+  * allowing the on-boarding of parties and uploading of packages. There are
+  * plans to make this functionality uniformly available: see the roadmap for
+  * progress information https://github.com/digital-asset/daml/issues/121.
+  *
+  * As of now there are five methods for changing the state of a DAML ledger:
+  * - submitting a transaction using [[WriteService!.submitTransaction]]
+  * - recording the rejection of a command submission using [[WriteService!.rejectSubmission]]
+  * - allocating a new party using [[WritePartyService!.allocateParty]]
+  * - uploading a new package using [[WritePackagesService!.uploadPackages]]
+  * - pruning a participant ledger using [[WriteParticipantPruningService!.prune]]
+  */
+trait WriteService
+    extends WritePackagesService
+    with WritePartyService
+    with WriteConfigService
+    with WriteParticipantPruningService
+    with ReportsHealth {
+
+  /** Submit a transaction for acceptance to the ledger.
+    *
+    * TODO(v2) Why do we need the requirement below on no IO? Canton doesn't respect it at the moment.
+    *  For command deduplication, we will need to do some IO inside the write service,
+    *  and want to report the outcome synchronously whenever possible.
```

Comment:
@andreaslochbihler-da : we can drop that requirement. It's a left-over from a time where the return result was not wrapped in a `CompletionStage`

Reply:
Great. I cut it down to "thread-safe".